examples/developer-tools/src/output-generators/factory_base.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as Blockly from 'blockly/core';
+import {javascriptGenerator} from 'blockly/javascript';
+import {
+  JsonDefinitionGenerator,
+  jsonDefinitionGenerator,
+  Order,
+} from './json_definition_generator';
+
+/**
+ * Builds the 'message0' part of the JSON block definition.
+ * The message should have label fields' text inlined into the message.
+ * Doing so makes the message more translatable as fields can be moved around.
+ *
+ * @param argsList The list of fields and inputs generated from the input stack.
+ * @returns An object containing:
+ *    - a message string with one placeholder '%i`
+ *      for each field and input in the block
+ *    - the new args list, with label fields removed
+ */
+const buildMessageString = function (argsList: Array<Record<string, unknown>>) {
+  let i = 0;
+  let messageString = '';
+  const newArgs = [];
+  for (const arg of argsList) {
+    if (arg.type === 'field_label') {
+      // Label fields get added directly to the message string.
+      // They are removed from the arg list so they don't appear twice.
+      messageString += `${arg.text} `;
+    } else {
+      i++;
+      messageString += `%${i} `;
+      newArgs.push(arg);
+    }
+  }
+
+  return {
+    message: messageString.trim(),
+    args: newArgs,
+  };
+};
+
+jsonDefinitionGenerator.forBlock['factory_base'] = function (
+  block: Blockly.Block,
+  generator: JsonDefinitionGenerator,
+): string {
+  // TODO: Get a JSON-legal name for the block
+  const blockName = block.getFieldValue('NAME');
+  // Tooltip and Helpurl string blocks can't be removed, so we don't care what happens if the block doesn't exist
+  const tooltip = JSON.parse(
+    generator.valueToCode(block, 'TOOLTIP', Order.ATOMIC),
+  );
+  const helpUrl = JSON.parse(
+    generator.valueToCode(block, 'HELPURL', Order.ATOMIC),
```

Comment:
I've heard of trailing commas on object literals, but trailing commas on function parameters is new to me.

Reply:
yeah, this was a relatively recent change by the formatter. I think they used the same logic for trailing commas elsewhere and applied it to multiline function definitions as well (smaller diffs when adding or removing parameters, etc).